src/RecursiveStagesService.php

```diff
@@ -0,0 +1,187 @@
+            if ($this->isModelDirty($model)) {
+                // Model is dirty,
+                // we can return here as there is no need to check the rest of the hierarchy
+                return true;
+            }
+
+            // Discover and add owned objects for inspection
+            $relatedObjects = $this->getOwnedObjects($model);
+            $models = array_merge($models, $relatedObjects);
+        }
+
+        // Compare deleted content,
+        // this wouldn't be covered in hierarchy traversal as deleted models are no longer present
+        $draftIdentifiers = $this->getOwnedIdentifiers($object, Versioned::DRAFT);
+        $liveIdentifiers = $this->getOwnedIdentifiers($object, Versioned::LIVE);
+
+        return $draftIdentifiers !== $liveIdentifiers;
+    }
+
+    /**
+     * Get unique identifiers for all owned objects, so we can easily compare them
+     *
+     * @param DataObject $object
+     * @param string $stage
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedIdentifiers(DataObject $object, string $stage): array
+    {
+        $identifiers = Versioned::withVersionedMode(function () use ($object, $stage): array {
+            Versioned::set_stage($stage);
+
+            /** @var DataObject $stagedObject */
+            $stagedObject = DataObject::get_by_id($object->ClassName, $object->ID);
+
+            if ($stagedObject === null) {
+                return [];
+            }
+
+            $models = [$stagedObject];
+            $identifiers = [];
+
+            while ($model = array_shift($models)) {
+                // Compose a unique identifier, so we can easily compare models
+                // Note that we intentionally use base class here, so we can cover the situation where model class changes
+                $identifiers[] = implode('_', [$model->baseClass(), $model->ID]);
+                $relatedObjects = $this->getOwnedObjects($model);
+                $models = array_merge($models, $relatedObjects);
+            }
+
+            return $identifiers;
+        });
+
+        sort($identifiers, SORT_STRING);
+
+        return array_values($identifiers);
+    }
+
+    /**
+     * This lookup will attempt to find "owned" objects
+     * This method uses the 'owns' relation, same as @see RecursivePublishable::publishRecursive()
+     *
+     * @param DataObject|RecursivePublishable $object
+     * @return array
+     * @throws Exception
+     */
+    protected function getOwnedObjects(DataObject $object): array
+    {
+        if (!$object->hasExtension(RecursivePublishable::class)) {
+            return [];
+        }
+
+        // Add versioned stage to cache key to cover the case where non-versioned model owns versioned models
+        // In this situation the versioned models can have different cached state which we need to cover
+        $cacheKey = sprintf('%s-%s', $object->getUniqueKey(), Versioned::get_stage());
```

Comment:
```suggestion
        $cacheKey = $object->getUniqueKey() . '-' . Versioned::get_stage();
```
This is easier to scan-read at a glance